### PR TITLE
[ValueTracking] Handle `icmp pred (trunc X), C` in `computeKnownBitsFromCmp`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch handles the pattern `icmp pred (trunc X), C` in `computeKnownBitsFromCmp` to infer low bits of `X` from dominating conditions.

### CODE REVIEW

File: llvm/lib/Analysis/DomConditionCache.cpp [hunk 1]
```diff
@@ -27,7 +27,8 @@ static void findAffectedValues(Value *Cond,
 
       // Peek through unary operators to find the source of the condition.
       Value *Op;
-      if (match(I, m_PtrToInt(m_Value(Op)))) {
+      if (match(I,
+                m_CombineOr(m_PtrToInt(m_Value(Op)), m_Trunc(m_Value(Op))))) {
```
Comment (nikic) [inline llvm/lib/Analysis/DomConditionCache.cpp:31]: We should also support for pattern for assumes.

Comment (goldsteinn) [reply llvm/lib/Analysis/DomConditionCache.cpp:31]: I'm working on a patch to unify the two findAffectedValues implementations.

Comment (dtcxzyw) [reply llvm/lib/Analysis/DomConditionCache.cpp]: Rebased on the top of https://github.com/llvm/llvm-project/pull/83161.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -724,10 +724,22 @@ static void computeKnownBitsFromCond(const Value *V, Value *Cond,
...
+    Value *LHS = Cmp->getOperand(0);
+    Value *RHS = Cmp->getOperand(1);
+
+    // Handle icmp pred (trunc V), C
+    if (match(LHS, m_Trunc(m_Specific(V)))) {
+      KnownBits DstKnown(LHS->getType()->getScalarSizeInBits());
+      computeKnownBitsFromCmp(LHS, Pred, LHS, RHS, DstKnown, SQ);
+      Known = Known.unionWith(DstKnown.anyext(Known.getBitWidth()));
+      return;
+    }
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:739]: Do you actually need seperate code for `m_Trunc`. Think the code you have in `m_Trunc` would would generically as well?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:739]: Could you please provide a test case for demonstration?

File: llvm/lib/Analysis/DomConditionCache.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Analysis/DomConditionCache.cpp]: Can you please also add a test using assume? I think to make it work you need to move the logic from computeKnownBitsFromCond into computeKnownBitsFromCmp.

Comment (dtcxzyw) [reply llvm/lib/Analysis/DomConditionCache.cpp]: Done.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp]: Done.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -815,8 +832,7 @@ void llvm::computeKnownBitsFromContext(const Value *V, KnownBits &Known,
     if (!isValidAssumeForContext(I, Q.CxtI, Q.DT))
       continue;
 
-    computeKnownBitsFromCmp(V, Cmp->getPredicate(), Cmp->getOperand(0),
-                            Cmp->getOperand(1), Known, Q);
+    computeKnownBitsFromICmpCond(V, Cmp, Known, Q, /*Invert=*/false);
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:835]: Thought the idea was move logic from `computeKnownBitsFromICmpCond` -> `computeKnownBitsFromCmp`, not use `computeKnownBitsFromICmpCond` for assumes.

**Comment by nikic** [comment]
Hm, I wonder why this causes code size regressions, especially on mafft (https://llvm-compile-time-tracker.com/compare.php?from=0cbbcf1ef006ce13a1fa94960067723982ae955a&to=3589cacfa8da89b9b5051e4dba659caa575e6b3f&stat=size-text).

**Comment by dtcxzyw** [comment]
> Hm, I wonder why this causes code size regressions, especially on mafft (https://llvm-compile-time-tracker.com/compare.php?from=0cbbcf1ef006ce13a1fa94960067723982ae955a&to=3589cacfa8da89b9b5051e4dba659caa575e6b3f&stat=size-text).

I will add CTMark to my benchmark.

**Comment by dtcxzyw** [comment]
> Hm, I wonder why this causes code size regressions, especially on mafft (https://llvm-compile-time-tracker.com/compare.php?from=0cbbcf1ef006ce13a1fa94960067723982ae955a&to=3589cacfa8da89b9b5051e4dba659caa575e6b3f&stat=size-text).

Reverting this patch cannot improve the code size.
https://github.com/dtcxzyw/llvm-ci/pull/1107#issuecomment-1983185593

**Pull request merged**